Datadog Connector Component

gord02 opened the pull request:
This pull request is for a newly created Datadog connector that replaces the need for the Datadog processor. This replacement is needed because the processor made use of a work-around that directly allowed the processor to export data. This is not something that adheres to the OpenTelemetry standard and thus the previous system should be deprecated and replaced by this new connector.

Link to issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19740

mx-psi commented on connector/datadogconnector/factory.go:
```diff
@@ -0,0 +1,41 @@
+	if err != nil {
+		return nil, err
+	}
+	c.metricsConsumer = nextConsumer
```
Instead of doing this, why not pass it as an argument to `newConnector`?

mx-psi commented on connector/datadogconnector/go.mod:
```diff
@@ -0,0 +1,88 @@
+go 1.20
+
+require (
+	github.com/DataDog/datadog-agent/pkg/proto v0.48.0-beta.1
```
@songy23 can you check what version should we use here and if we need replace statements?

songy23 replied:
We should use `v0.48.0-beta.1` , if the new component is included in dependabot config then we need replace as well.

mx-psi commented on connector/datadogconnector/go.mod:
```diff
@@ -0,0 +1,88 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector
+
+go 1.20
```
We are still using Go 1.19, this will be changed when #25116 is addressed

```suggestion
go 1.19
```

mx-psi commented on go.mod:
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/clickhouseexporter v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/coralogixexporter v0.82.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter v0.82.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter v0.0.0-00010101000000-000000000000
```
You can revert this change, it should stay as it is

mx-psi commented on go.mod:
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/attributesprocessor v0.82.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor v0.82.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/processor/datadogprocessor v0.82.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/datadogprocessor v0.0.0-00010101000000-000000000000
```
ditto, revert this change

songy23 commented on connector/datadogconnector/go.mod:
```diff
@@ -0,0 +1,88 @@
+	google.golang.org/protobuf v1.31.0 // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+)
+
```
Please add these 2 lines:
```
replace github.com/DataDog/datadog-agent/pkg/proto => github.com/DataDog/datadog-agent/pkg/proto v0.48.0-beta.1

replace github.com/DataDog/datadog-agent/pkg/trace => github.com/DataDog/datadog-agent/pkg/trace v0.48.0-beta.1
```

songy23 reviewed (commented):
Is this ready for review? 

Also you need a rebase

gord02 committed:
parent 0454100f244052c42b968ad4c15313474a49e184
author Gordon Hamilton <gordon.hamilton@datadoghq.com> 1691699624 -0400
committer Gordon Hamilton <gordon.hamilton@datadoghq.com> 1691701555 -0400

cleaning up commit

gord02 committed:
resolving conflicts

mx-psi commented on connector/datadogconnector/metadata.yaml:
```diff
@@ -0,0 +1,9 @@
+  class: connector
+  stability:
+    alpha: [traces_to_metrics]
+  distributions: [contrib]
```
For now, this is not on the contrib distribution, let's leave this for after we make a PR on opentelemetry-collector-releases and actually add this to the contrib distro.
```suggestion
  distributions: []
```

songy23 replied:
This is not resolved.

mx-psi commented on connector/datadogconnector/connector.go:
```diff
@@ -0,0 +1,115 @@
+type connectorImp struct {
+	metricsConsumer consumer.Metrics // the next component in the pipeline to ingest data after connector
+	logger          *zap.Logger
+	started         bool
```
We don't actually read this value, only set it, so we can get rid of it
```suggestion
```

(and also remove other references below)

mx-psi commented on connector/datadogconnector/connector_test.go:
```diff
@@ -0,0 +1,32 @@
+	traceConnector, err := factory.CreateTracesToMetrics(context.Background(), creationParams, cfg, consumertest.NewNop())
+	smc := traceConnector.(*connectorImp)
+
+	assert.Nil(t, err)
```
nit: the usual way of doing this is 
```suggestion
	assert.NoError(t, err)
```

this gives a nicer error message

mx-psi commented on connector/datadogconnector/connector_test.go:
```diff
@@ -0,0 +1,32 @@
+
+	// Test
+	traceConnector, err := factory.CreateTracesToMetrics(context.Background(), creationParams, cfg, consumertest.NewNop())
+	smc := traceConnector.(*connectorImp)
```
This will panic if `traceConnector` is not of that type, see https://go.dev/tour/methods/15. Instead of doing this, use the `ok` form, and check that it returned true:
```suggestion
	smc, ok := traceConnector.(*connectorImp)
	assert.True(t, ok)
```

gord02 committed:
resolving issues to pass CI checks

gord02 committed:
adding changelog file

songy23 commented on connector/datadogconnector/connector_test.go:
```diff
@@ -0,0 +1,31 @@
+	"go.opentelemetry.io/collector/consumer/consumertest"
+)
+
+var _ component.Component = (*connectorImp)(nil) // testing that the connectorImp properly implements the type Component interface
```
This should be in `connector.go` instead

songy23 reviewed (commented):
Please follow https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components, you're missing several steps like add to versions.yaml, add doc.go, make generate-gh-issue-templates, etc.

gord02 committed:
updating verison.yaml

gord02 committed:
adding in doc.go and other changes necessary for adding new component

songy23 committed:
Merge branch 'main' into datadogconnector

songy23 commented on versions.yaml:
```diff
@@ -16,6 +16,7 @@ module-sets:
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector
       - github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector
+      - github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector
```
Please keep in alphabetic order.

songy23 commented on cmd/otelcontribcol/builder-config.yaml:
```diff
@@ -201,6 +201,7 @@ connectors:
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector v0.82.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector v0.82.0
   - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector v0.82.0
+  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector v0.82.0
```
Please keep in alphabetic order.

songy23 commented on cmd/otelcontribcol/builder-config.yaml:
```diff
@@ -416,6 +417,7 @@ replaces:
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector => ../../connector/routingconnector
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector => ../../connector/servicegraphconnector
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector => ../../connector/spanmetricsconnector
+  - github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector => ../../connector/datadogconnector
```
same

gord02 committed:
fixing code based on feedback

gord02 committed:
Merge branch 'datadogconnector' of github.com:gord02/opentelemetry-collector-contrib into datadogconnector

gord02 committed:
re-running crosslink

songy23 commented:
> Generated code is out of date, please run "make genotelcontribcol" and commit the changes in this PR.

gord02 committed:
make genotelcontribcol

mx-psi commented on .chloggen/datadogconnector.yaml:
```diff
@@ -0,0 +1,27 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
```suggestion
subtext: | 
  This component is meant to replace the Datadog processor
```

mx-psi committed:
Update .chloggen/datadogconnector.yaml

mx-psi committed:
Merge branch 'main' into datadogconnector
